File: GithubUser/GithubUser/Model/UserDTO.swift
//
//  UserDTO.swift
//  GithubUser
//
//  Created by Lam Le on 15/06/2022.
//

import Foundation

/// Requirement:
/// Display username, github profile’s link and avatar
struct UserDTO: Codable {

    var userName: String = ""
    var githubProfile: String = ""
    var avatar: String = ""
    var location: String = ""
    var bio: String = ""
    var publicRepos: Int = 0
    var followers: Int = 0
    var following: Int = 0

    enum CodingKeys: String, CodingKey {
        case userName = "login"
        case githubProfile = "url"
        case avatar = "avatar_url"
        case location = "location"
        case bio = "bio"
        case publicRepos = "public_repos"
        case followers = "followers"
        case following = "following"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        userName <- container[.userName]
        githubProfile <- container[.githubProfile]
        avatar <- container[.avatar]
        location <- container[.location]
        bio <- container[.bio]
        publicRepos <- container[.publicRepos]
        followers <- container[.followers]
        following <- container[.following]
    }
}

File: fl_github_users/lib/model/user_dto.dart

class UserDTO {
  final int id;
  final String login;
  final String githubProfile;
  final String avatar;

  const UserDTO({
    required this.id,
    required this.login,
    required this.githubProfile,
    required this.avatar
  });

  factory UserDTO.fromJSON(Map<String, dynamic> json) {
    return UserDTO(
        id: json["id"],
        login: json["login"],
        githubProfile: json["url"],
        avatar: json["avatar_url"]);
  }
}
